File: worker-supertonic/redis_messages.py
"""Typed Redis Stream messages shared with the QueueTts Gateway contract.

The gateway serializes nested payload/result values as JSON strings while the
outer message is a flat Redis Stream field map.  Models in this module keep
that wire format at the Redis boundary and expose typed Python objects to the
worker.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime, timezone
from enum import StrEnum
from typing import Any, Mapping, Protocol, TypeAlias


JsonObject: TypeAlias = dict[str, Any]


class RedisModelError(ValueError):
    """Raised when a Redis message does not satisfy the gateway contract."""


class JsonModel(Protocol):
    def to_dict(self) -> JsonObject: ...


class JobType(StrEnum):
    TTS = "tts"
    STYLES = "styles"


class ResultState(StrEnum):
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"


def _json_object(value: Any, field_name: str) -> JsonObject:
    if value is None or value == "":
        return {}
    if isinstance(value, str):
        try:
            value = json.loads(value)
        except json.JSONDecodeError as exc:
            raise RedisModelError(f"'{field_name}' is not valid JSON: {exc.msg}") from exc
    if not isinstance(value, Mapping):
        raise RedisModelError(f"'{field_name}' must be a JSON object")
    return dict(value)


def _required_text(data: Mapping[str, Any], name: str) -> str:
    value = data.get(name)
    if value is None:
        raise RedisModelError(f"missing '{name}' in payload")
    return str(value)


def _optional_text(value: Any) -> str | None:
    return None if value is None else str(value)


def _optional_int(value: Any) -> int | None:
    return None if value is None else int(value)


def _optional_float(value: Any) -> float | None:
    return None if value is None else float(value)


@dataclass(frozen=True, slots=True)
class TtsPayload:
    text: str
    voice: str | None = None
    lang: str | None = None
    speed: float | None = None
    steps: int | None = None
    response_format: str | None = None
    seed: int | None = None
    max_chunk_length: int | None = None
    silence_duration: float | None = None

    @classmethod
    def from_dict(cls, data: Mapping[str, Any]) -> "TtsPayload":
        return cls(
            text=_required_text(data, "text"),
            voice=_optional_text(data.get("voice")),
            lang=_optional_text(data.get("lang")),
            speed=_optional_float(data.get("speed")),
            steps=_optional_int(data.get("steps")),
            response_format=_optional_text(data.get("response_format")),
            seed=_optional_int(data.get("seed")),
            max_chunk_length=_optional_int(data.get("max_chunk_length")),
            silence_duration=_optional_float(data.get("silence_duration")),
        )


@dataclass(frozen=True, slots=True)
class EmptyPayload:
    @classmethod
    def from_dict(cls, data: Mapping[str, Any]) -> "EmptyPayload":
        return cls()


JobPayload: TypeAlias = TtsPayload | EmptyPayload


def parse_job_payload(job_type: JobType, data: Mapping[str, Any]) -> JobPayload:
    model = {
        JobType.TTS: TtsPayload,
        JobType.STYLES: EmptyPayload,
    }[job_type]
    return model.from_dict(data)


@dataclass(frozen=True, slots=True)
class RedisJobMessage:
    job_id: str
    job_type: JobType
    payload: JobPayload

    @classmethod
    def from_redis_fields(cls, fields: Mapping[str, Any]) -> "RedisJobMessage":
        job_id = str(fields.get("jobId") or "").strip()
        if not job_id:
            raise RedisModelError("Redis job message has no jobId")
        type_value = str(fields.get("type") or JobType.TTS)
        try:
            job_type = JobType(type_value)
        except ValueError as exc:
            raise RedisModelError(f"unsupported job type: {type_value}") from exc
        payload_data = _json_object(fields.get("payload"), "payload")
        return cls(
            job_id=job_id,
            job_type=job_type,
            payload=parse_job_payload(job_type, payload_data),
        )


@dataclass(frozen=True, slots=True)
class AudioResult:
    duration_s: float
    sample_rate: int
    media_type: str
    audio_format: str
    processed_text: str

    def to_dict(self) -> JsonObject:
        return {
            "durationS": self.duration_s,
            "sampleRate": str(self.sample_rate),
            "mediaType": self.media_type,
            "audioFormat": self.audio_format,
            "processedText": self.processed_text,
        }


@dataclass(frozen=True, slots=True)
class ArtifactContent:
    file_name: str
    media_type: str
    content_base64: str

    def to_dict(self) -> JsonObject:
        return {
            "fileName": self.file_name,
            "mediaType": self.media_type,
            "contentBase64": self.content_base64,
        }


@dataclass(frozen=True, slots=True)
class StyleInfo:
    name: str
    kind: str
    path: str | None = None

    def to_dict(self) -> JsonObject:
        return _without_none({
            "name": self.name,
            "kind": self.kind,
            "path": self.path,
        })


@dataclass(frozen=True, slots=True)
class StylesResult:
    styles: list[StyleInfo]
    worker_id: str
    batch_size: int

    def to_dict(self) -> JsonObject:
        return {
            "styles": [style.to_dict() for style in self.styles],
            "worker_id": self.worker_id,
            "batch_size": self.batch_size,
        }


@dataclass(frozen=True, slots=True)
class ErrorResult:
    message: str
    code: str | None = None
    detail: str | None = None
    reason: str | None = None

    def to_dict(self) -> JsonObject:
        return _without_none({
            "code": self.code,
            "message": self.message,
            "detail": self.detail,
            "reason": self.reason,
        })


ResultPayload: TypeAlias = AudioResult | StylesResult


@dataclass(frozen=True, slots=True)
class RedisResultMessage:
    job_id: str
    worker_id: str
    batch_id: str
    state: ResultState
    result: ResultPayload | None = None
    error: ErrorResult | None = None
    artifact: ArtifactContent | None = None
    started_at: datetime | None = None

    @classmethod
    def running(
        cls,
        *,
        job_id: str,
        worker_id: str,
        batch_id: str,
        started_at: datetime | None = None,
    ) -> "RedisResultMessage":
        return cls(
            job_id,
            worker_id,
            batch_id,
            ResultState.RUNNING,
            started_at=started_at or datetime.now(timezone.utc),
        )

    @classmethod
    def succeeded(
        cls,
        *,
        job_id: str,
        worker_id: str,
        batch_id: str,
        result: ResultPayload,
        artifact: ArtifactContent | None = None,
    ) -> "RedisResultMessage":
        return cls(job_id, worker_id, batch_id, ResultState.SUCCEEDED, result=result, artifact=artifact)

    @classmethod
    def failed(
        cls,
        *,
        job_id: str,
        worker_id: str,
        batch_id: str,
        message: str,
    ) -> "RedisResultMessage":
        return cls(job_id, worker_id, batch_id, ResultState.FAILED, error=ErrorResult(message))

    def to_redis_fields(self) -> dict[str, str]:
        fields = {
            "jobId": self.job_id,
            "workerId": self.worker_id,
            "batchId": self.batch_id,
            "state": self.state.value,
        }
        if self.result is not None:
            fields["result"] = _json_dump(self.result)
        if self.error is not None:
            fields["error"] = _json_dump(self.error)
        if self.artifact is not None:
            fields["artifact"] = _json_dump(self.artifact)
        if self.started_at is not None:
            fields["startedAt"] = self.started_at.isoformat()
        return fields


def _without_none(data: Mapping[str, Any]) -> JsonObject:
    return {key: value for key, value in data.items() if value is not None}


def _json_dump(value: JsonModel | Mapping[str, Any]) -> str:
    data = value.to_dict() if hasattr(value, "to_dict") else dict(value)
    return json.dumps(data, ensure_ascii=False, separators=(",", ":"))

File: gateway-api/src/main/kotlin/com/aitts/queuetts/gateway/api/infra/redis/messaging/RedisMessages.kt
package com.aitts.queuetts.gateway.api.infra.redis.messaging

import com.aitts.queuetts.gateway.api.dto.AudioJobResult
import com.aitts.queuetts.gateway.api.dto.JobError
import com.aitts.queuetts.gateway.api.dto.JobPayload
import com.fasterxml.jackson.annotation.JsonAlias
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.ObjectMapper
import java.time.OffsetDateTime

// Redis Streams 로 주고받는 메시지 계약. 워커의 redis_messages.py 와 대칭을 이룬다.
// (보냄: 우선순위별 job 스트림 / 받음: result 스트림)

/**
 * 게이트웨이가 우선순위별 job 스트림에 XADD 하는 발행 메시지.
 *
 * payload 는 tts 잡이면 [JobPayload], `styles` 처럼 인자가 없는 제어 요청이면 비어 있고,
 * `speaker_forget` 처럼 인자가 있는 제어 요청이면 그 요청의 DTO 다. 어차피 JSON 문자열
 * 한 필드로 나가므로 타입을 하나로 묶지 않는다.
 */
data class RedisJobMessage(
    val jobId: String,
    val type: String,
    val payload: Any?,
    val priority: String,
    val source: String?,
    val enqueuedAt: OffsetDateTime,
) {
    /**
     * Redis Stream 필드 맵으로 직렬화한다. 바깥은 평면 필드, payload 만 JSON 문자열이다.
     *
     * payload 가 [com.aitts.queuetts.gateway.api.dto.JobPayload] 여도 **worker 가 보는 JSON 은 예전
     * 그대로 평면**이다 — 모델별 타입은 Gateway 안의 구분이고, 각 구현체가 자기 엔진의 필드만
     * NON_NULL 로 내보내므로 worker 는 자기 파라미터만 담긴 객체를 받는다. 이 경계 덕분에 이
     * 구조 변경으로 워커를 고치지 않는다.
     */
    fun toRedisFields(objectMapper: ObjectMapper): Map<String, String> = buildMap {
        put("jobId", jobId)
        put("type", type)
        put("payload", objectMapper.writeValueAsString(payload ?: emptyMap<String, Any?>()))
        put("priority", priority)
        source?.let { put("source", it) }
        put("enqueuedAt", enqueuedAt.toString())
    }
}

/** worker 가 result 스트림으로 돌려주는 진행/완료 메시지. */
data class RedisResultMessage(
    @field:JsonAlias("job_id")
    val jobId: String,
    @field:JsonAlias("worker_id")
    val workerId: String = "unknown-worker",
    @field:JsonAlias("batch_id")
    val batchId: String? = null,
    @field:JsonAlias("status")
    val state: String? = null,
    @field:JsonAlias("started_at")
    val startedAt: OffsetDateTime? = null,
    val result: AudioJobResult? = null,
    val styleCatalog: StyleCatalogResult? = null,
    /**
     * 제어 요청 응답의 `result` JSON 원문.
     *
     * 제어 요청마다 응답 모양이 다르므로(styles / speaker_forget 등) 여기서 타입을 정하지 않고
     * 요청을 보낸 쪽이 자기가 아는 타입으로 읽는다. tts 잡에서는 null 이다.
     */
    val rawResult: String? = null,
    val error: JobError? = null,
    val artifact: ArtifactContent? = null,
)

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
data class StylesCatalogVoice(
    val name: String,
    val kind: String,
    val path: String? = null,
)

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
data class StyleCatalogResult(
    @field:JsonAlias("styles")
    val styles: List<StylesCatalogVoice> = emptyList(),
    @field:JsonProperty("worker_id")
    @field:JsonAlias("workerId")
    val workerId: String? = null,
    // worker 가 styles 응답에 함께 실어 보내는 자기 batchSize (표시용). worker 소유 값.
    @field:JsonAlias("batch_size")
    val batchSize: Int? = null,
)

/**
 * `speaker_forget` 응답. 등록과 참조 변경은 Gateway 단독으로 처리하므로 prepare 응답은 없다.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
data class SpeakerControlResult(
    @field:JsonAlias("speaker_name", "speakerId", "speaker_id", "styleId", "style_id")
    val speakerName: String = "",
    /** 지울 캐시가 있었는지. */
    val applied: Boolean? = null,
    @field:JsonProperty("worker_id")
    @field:JsonAlias("workerId")
    val workerId: String? = null,
)

/** result 메시지에 실려 오는 생성 아티팩트(오디오) 콘텐츠. */
data class ArtifactContent(
    @field:JsonAlias("file_name")
    val fileName: String? = null,
    @field:JsonAlias("media_type")
    val mediaType: String? = null,
    @field:JsonAlias("content_base64")
    val contentBase64: String? = null,
)
